src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -544,24 +554,123 @@ void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo
+				MarkDynamicDependency (dynamicDependency, (MemberReference) provider);
 		}
 
 		protected virtual bool ProcessLinkerSpecialAttribute (CustomAttribute ca, ICustomAttributeProvider provider, in DependencyInfo reason)
 		{
-			if (IsUserDependencyMarker (ca.AttributeType) && provider is MemberReference mr) {
+			var isPreserveDependency = IsUserDependencyMarker (ca.AttributeType);
+			var isDynamicDependency = LinkerAttributesInformation.IsAttribute<DynamicDependencyAttribute> (ca.AttributeType);
+
+			if (!((isPreserveDependency || isDynamicDependency) && provider is MemberReference mr))
+				return false;
+
+#if !FEATURE_ILLINK
+			if (isPreserveDependency)
 				MarkUserDependency (mr, ca);
+#endif
 
-				if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
-					MarkCustomAttribute (ca, reason);
-				} else {
-					// Record the custom attribute so that it has a reason, without actually marking it.
-					Tracer.AddDirectDependency (ca, reason, marked: false);
+			if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
+				MarkCustomAttribute (ca, reason);
+			} else {
+				// Record the custom attribute so that it has a reason, without actually marking it.
+				Tracer.AddDirectDependency (ca, reason, marked: false);
+			}
+
+			return true;
+		}
+
+		void MarkDynamicDependency (DynamicDependency dynamicDependency, MemberReference context)
+		{
+			Debug.Assert (context is MethodDefinition || context is FieldDefinition);
+			AssemblyDefinition assembly;
+			if (dynamicDependency.AssemblyName is string assemblyName) {
+				assembly = _context.GetLoadedAssembly (assemblyName);
+				if (assembly == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved assembly '{assemblyName}' in DynamicDependencyAttribute on '{context}'", 2031));
+					return;
 				}
+			} else {
+				assembly = context.Module.Assembly;
+				Debug.Assert (assembly != null);
+			}
 
-				return true;
+			TypeDefinition type;
+			if (dynamicDependency.TypeName is string typeName) {
+				type = assembly.FindTypeByDocumentationSignature (typeName);
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{typeName}' in DynamicDependencyAttribute on '{context}'", 2032));
+					return;
+				}
+			} else if (dynamicDependency.Type is TypeReference typeReference) {
+				type = typeReference.Resolve ();
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{typeReference}' in DynamicDependencyAtribute on '{context}'", 2032));
+					return;
+				}
+			} else {
+				type = context.DeclaringType.Resolve ();
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{context.DeclaringType}' in DynamicDependencyAttribute on '{context}'", 2032));
+					return;
+				}
```

Comment:
I'm tempted for this logic to respect `--skip-unresolved`

Reply:
That is good point. Curious: do normal apps run linker with `--skip-unresolved` or is that more of a corner case option?

Reply:
It's not a default option it's more about avoiding the aggressive context loading again

Reply:
We do pass it when running the linker from the SDK. As I understand it, that option enables the linker to continue from resolution failures (see https://github.com/mono/linker/pull/260) instead of crashing. I don't think we need to respect it here since these warnings won't fail the build. I think that if `--skip-unresolved` is false, the call to `Resolve()` would cause an `AssemblyResolutionException` and we wouldn't even get to log the warning.